File: bun-budget-api/src/services/report-service.ts
import { Transaction as ProcessedTransaction } from "../../prisma/client";
import prisma from "../utils/prisma";

const roundToTwoDecimals = (num: number) => Math.round(num * 100) / 100;

const initialReport = {
  transactions_sum: 0,
  expenses_sum: 0,
  incomes_sum: 0,
  transactions_count: 0,
  expenses_count: 0,
  incomes_count: 0,
};

const mapReport = (transactions: ProcessedTransaction[]) =>
  transactions.reduce((report, transaction) => {
    const value = Number(transaction.value);
    const isExpense = value < 0;
    const isIncome = value > 0;

    return {
      transactions_sum: roundToTwoDecimals(report.transactions_sum + value),
      expenses_sum: isExpense
        ? roundToTwoDecimals(report.expenses_sum + value)
        : report.expenses_sum,
      incomes_sum: isIncome
        ? roundToTwoDecimals(report.incomes_sum + value)
        : report.incomes_sum,
      transactions_count: report.transactions_count + 1,
      expenses_count: isExpense
        ? report.expenses_count + 1
        : report.expenses_count,
      incomes_count: isIncome ? report.incomes_count + 1 : report.incomes_count,
    };
  }, initialReport);

export const generateReport = async (
  credentialId: string,
  category?: string,
  from?: Date,
  to?: Date,
) => {
  const transactions = await prisma.transaction.findMany({
    where: { credentialId, category, timestamp: { gte: from, lte: to } },
  });

  const report = mapReport(transactions);

  return report;
};

File: cs-budget-api/main/src/Services/ReportService.cs
using Data;
using Microsoft.EntityFrameworkCore;
using Models;

namespace Services;

public class ReportService(DatabaseContext dbContext)
{

    private static BudgetReport MapBudgetReport(List<decimal> transactionValues)
    {
        var budgetReport = transactionValues.Aggregate(new BudgetReport(0, 0, 0, 0, 0, 0), (incompleteReport, value) =>
        {
            var isExpense = value < 0;
            var isIncome = value > 0;

            return new BudgetReport(
                transactions_sum: incompleteReport.transactions_sum + value,
                transactions_count: incompleteReport.transactions_count + 1,
                incomes_sum: isIncome ? incompleteReport.incomes_sum + value : incompleteReport.incomes_sum,
                incomes_count: isIncome ? incompleteReport.incomes_count + 1 : incompleteReport.incomes_count,
                expenses_sum: isExpense ? incompleteReport.expenses_sum + value : incompleteReport.expenses_sum,
                expenses_count: isExpense ? incompleteReport.expenses_count + 1 : incompleteReport.expenses_count
            );
        });

        return budgetReport;
    }

    public async Task<BudgetReport> GenerateReportAsync(Guid credentialId, string? category, DateTimeOffset? to, DateTimeOffset? from)
    {
        var transactionValues = await dbContext.Transactions
            .Where(transaction => transaction.CredentialId == credentialId
                && (category == null || transaction.Category == category)
                && (from == null || transaction.Timestamp >= from)
                && (to == null || transaction.Timestamp <= to))
            .Select(transaction => transaction.Value)
            .ToListAsync();

        var budgetReport = MapBudgetReport(transactionValues);

        return budgetReport;
    }
}
